File: Crio/crio_programming_interview_problems-master/PathsWithSumRootToLeaf/PathsWithSumRootToLeaf.cpp
#include<bits/stdc++.h>
#include "../crio/cpp/ds/TreeNode/TreeNode.hpp"
using namespace std;
#include "Solution.cpp"

// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

TreeNode *makeTreeFromEdges(vector<long> val , vector<pair<long, long> > edges)
{
    unordered_map<int, TreeNode*> nodebyIndex;
    int n = val.size();
    for (long i = 0; i < n; i++)
        nodebyIndex[i] = new TreeNode(val[i]);
    TreeNode *root = nodebyIndex[0];
    for (int i = 0; i < n; i++)
    {
        TreeNode *ptr = nodebyIndex[i];
        if (edges[i].first != -1) {

            ptr->left = nodebyIndex[edges[i].first];
        }
        if (edges[i].second != -1) {
            ptr->right = nodebyIndex[edges[i].second];
        }
    }
    return root;
}
int main()
{
    int tests;
    cin >> tests;
    while (tests--)
    {
        int n, a, b;
        cin >> n;
        vector<long > val(n, 0);
        for (int i = 0; i < n; i++)
            cin >> val[i];
        vector<pair<long, long> > edges;
        for ( int i = 0; i < n; i++) {
            int x;
            cin >> x; cin >> a >> b;
            edges.push_back({a-1, b-1});
        }
        TreeNode *root = makeTreeFromEdges(val, edges);
        long long sum;
        cin >> sum;
        long long answer = Solution().pathsWithSumRootToLeaf(root,sum);
        cout << answer << endl;
    }
}

File: Crio/crio_programming_interview_problems-master/PathsWithSumRootToLeaf/PathsWithSumRootToLeaf.java
import java.util.*;
import crio.ds.Tree.TreeNode;

// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

class Pair {
    int first;
    int second;
    Pair(int a, int b) {
        first = a;
        second = b;
    }
}

public class PathsWithSumRootToLeaf {
    TreeNode makeTreeFromEdges(long val[] , ArrayList<Pair> edges) {
        HashMap<Integer , TreeNode> nodebyIndex = new HashMap<Integer , TreeNode>();
        int n = val.length;
        for (int i = 0; i < n; i++) {
            nodebyIndex.put(i, new TreeNode(val[i]));
        }
        TreeNode root = nodebyIndex.get(0);
        for (int i = 0; i < n; i++) {
            TreeNode ptr = nodebyIndex.get(i);
            if (edges.get(i).first != -1) {

                ptr.left = nodebyIndex.get(edges.get(i).first);
            }
            if (edges.get(i).second != -1) {
                ptr.right = nodebyIndex.get(edges.get(i).second);
            }
        }
        return root;
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int tests = scanner.nextInt();
        for(int t = 0; t <  tests ; t++) {
            int n, a, b;
            n = scanner.nextInt();
            long val[] = new long[n];
            for (int i = 0; i < n; i++) {
                val[i] = scanner.nextLong();
            }
            ArrayList<Pair> edges = new ArrayList<Pair>();
            for ( int i = 0; i < n; i++) {
                int x;
                x = scanner.nextInt(); 
                a = scanner.nextInt();
                b = scanner.nextInt();
                edges.add(new Pair(a-1, b-1));
            }
            TreeNode root = new PathsWithSumRootToLeaf().makeTreeFromEdges(val, edges);
            long sum;
            sum = scanner.nextLong();
            long answer = new Solution().pathsWithSumRootToLeaf(root,sum);
            System.out.println(answer);
        }
    }
    
}
